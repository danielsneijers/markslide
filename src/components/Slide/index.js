// @flow
import React from 'react'
import CSSTransitionGroup from 'react-transition-group/CSSTransitionGroup'
import CSS from './style.css'

type Props = {
  key?: string,
  content: string,
  onClick: Function
}

const Slide = ({ content, ...rest }: Props) => {
  return (
    <CSSTransitionGroup
      className={CSS.slide}
      transitionName='slide'
      transitionAppearTimeout={500}
      transitionEnterTimeout={500}
      transitionLeaveTimeout={300}
      transitionAppear
      component='div'
      {...rest}
    >
      <div dangerouslySetInnerHTML={{ __html: content }} />
    </CSSTransitionGroup>
  )
}

export default Slide
