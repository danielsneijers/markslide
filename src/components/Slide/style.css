.slide {
  height: 100vh;
  width: 100vw;
  background-color: #fafafa;
  padding: 10%;
  display: flex;
  align-items: center;
  justify-content: center;
}

:global .slide-appear {
  opacity: 0.01;
}

:global .slide-appear.slide-appear-active {
  opacity: 1;
  transition: opacity .5s ease-in;
}

:global .slide-enter {
  opacity: 0.01;
}

:global .slide-enter.slide-enter-active {
  opacity: 1;
  transition: opacity 500ms ease-in;
}

:global .slide-leave {
  opacity: 1;
}

:global .slide-leave.slide-leave-active {
  opacity: 0.01;
  transition: opacity 300ms ease-in;
}
