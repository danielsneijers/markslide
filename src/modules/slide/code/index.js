export { default as CodeSlide } from './component'
export { default as CodeSlideContainer } from './container'
