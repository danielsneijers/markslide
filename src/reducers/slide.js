export default function slides (state = [], action) {
  const { type } = action

  switch (type) {
    default:
      return state
  }
}
