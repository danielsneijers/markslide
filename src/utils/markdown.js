// @flow
import Marked from 'marked'
import { compose } from 'ramda'
import { separateContentAndMetaData } from 'utils/metaData'
import type { SlideWithMetaData } from 'utils/metaData'
import renderer from 'utils/renderer'

type ParsedSlideWithMetaData = {
  content: HTMLElement,
  meta: {}
}

export const splitSlides = (content: string): Array<string> =>
  content.split('\n---\n')

export const slidesCount = (content: string): number =>
  splitSlides(content).length

export const parseSlidesToEnhancedObjects = (
  slides: Array<string>
): Array<SlideWithMetaData> => slides.map(separateContentAndMetaData)

export const parseSlideObjectsToHtml = (
  slides: Array<SlideWithMetaData>
): Array<ParsedSlideWithMetaData> =>
  slides.map(({ content, meta }: SlideWithMetaData) => {
    return { content: Marked(content, { renderer }), meta }
  })

export const parse: Function = compose(
  parseSlideObjectsToHtml,
  parseSlidesToEnhancedObjects,
  splitSlides
)
