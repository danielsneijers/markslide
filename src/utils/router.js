import { get } from 'lodash'

export const getFromRouterProps = (props, param) => {
  return get(props, `match.params.${param}`)
}

export const getParamFromRouterProps = (param) => (props) => {
  return getFromRouterProps(props, param)
}

export const getSlideIndexFromProps = getParamFromRouterProps('slide')
