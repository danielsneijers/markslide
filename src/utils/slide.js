// @flow
import React from 'react'
import Slide from 'components/Slide'
import SlideCode from 'components/SlideCode'

type SlideProps = {
  key: string,
  onClick: Function
}

export const getSlideType = (content: string): string => {
  if (content.includes('<pre><code')) {
    return 'code'
  }

  return 'default'
}

export const getSlide = (content: string, elementProps: SlideProps): React$Element<any> => {
  const { key, onClick } = elementProps

  switch (getSlideType(content)) {
    case 'code':
      return <SlideCode key={key} content={content} onClick={onClick} />
    default:
      return <Slide key={key} content={content} onClick={onClick} />
  }
}
