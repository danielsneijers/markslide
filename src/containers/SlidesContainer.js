// @flow
import React, { PureComponent } from 'react'
import { throttle } from 'lodash'
import { withRouter } from 'react-router'
import { KEY_CODES } from 'constants'
import MarkdownSheets from 'config/slides.md'
import { parse, slidesCount } from 'utils/markdown'
import { getSlideIndexFromProps } from 'utils/router'
import { getProgressBar } from 'utils/settings'
import { getSlide } from 'utils/slide'

type Props = {
  navigateTo: Function
}

const NEXT = 'next'
const PREV = 'prev'

class SlidesContainer extends PureComponent {
  props: Props
  totalSlides: number
  currentIndex: number
  container: HTMLDivElement

  componentWillMount () {
    this.totalSlides = slidesCount(MarkdownSheets)
    this.updateIndex(this.props)
  }

  componentDidMount () {
    this.container.addEventListener('click', this.handleClick)
    window.addEventListener('keydown', this.handleKeyDown)
  }

  componentWillReceiveProps (nextProps: Props) {
    this.updateIndex(nextProps)
  }

  componentWillUnMount () {
    this.container.removeEventListener('click', this.handleClick)
    window.removeEventListener('keydown', this.handleKeyDown)
  }

  updateIndex (props: Props) {
    this.currentIndex = parseInt(getSlideIndexFromProps(props))

    if (this.currentIndex > this.totalSlides) {
      props.navigateTo(`/${this.totalSlides}`)
    }

    if (this.currentIndex < 1) {
      props.navigateTo('/1')
    }
  }

  handleClick = (e: MouseEvent) => {
    e.preventDefault()
    this.navigate(NEXT)
  }

  handleKeyDown = (e: KeyboardEvent) => {
    switch (e.keyCode) {
      case KEY_CODES.LEFT:
      case KEY_CODES.UP:
        this.navigate(PREV)
        break
      case KEY_CODES.RIGHT:
      case KEY_CODES.DOWN:
        this.navigate(NEXT)
        break
      case KEY_CODES.F:
        this.enterFullScreen(this.container)
        break
    }
  }

  navigate = throttle((direction) => {
    direction === PREV
      ? this.currentIndex > 1 && this.props.navigateTo(`/${this.currentIndex - 1}`)
      : this.currentIndex < this.totalSlides && this.props.navigateTo(`/${this.currentIndex + 1}`)
  }, 200)

  enterFullScreen (elem: HTMLDivElement) {
    if (elem.requestFullscreen) elem.requestFullscreen()
    // $FlowFixMe
    if (elem.webkitRequestFullscreen) elem.webkitRequestFullscreen()
    // $FlowFixMe
    if (elem.mozRequestFullScreen) elem.mozRequestFullScreen()
    // $FlowFixMe
    if (elem.msRequestFullscreen) elem.msRequestFullscreen()
  }

  render () {
    const parsedMarkdown = parse(MarkdownSheets)
    const progressBarProps = { offset: (1 - this.currentIndex / this.totalSlides) * 100 }
    const { content = '' } = parsedMarkdown[this.currentIndex - 1] || {}
    const slideProps = {
      uniqueKey: `slide-${this.currentIndex}`
    }

    return (
      <div ref={(c) => { this.container = c }}>
        {getProgressBar(progressBarProps)}
        {getSlide(content, slideProps)}
      </div>
    )
  }
}

export default withRouter(SlidesContainer)
export const withOutRouter = SlidesContainer
