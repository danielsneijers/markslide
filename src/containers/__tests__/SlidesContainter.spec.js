import React from 'react'
import { shallow } from 'enzyme'
import mockSheets from 'config/__mocks__/slides'
import * as MarkdownUtils from 'utils/markdown'
import * as SlideUtils from 'utils/slide'
import * as RouterUtils from 'utils/router'
import Slide from 'components/Slide'
import { withOutRouter as SlidesContainer } from '../SlidesContainer'

let props, tree

describe('containers/SlidesContainer', () => {
  beforeEach(() => {
    const content = MarkdownUtils.parse(mockSheets)

    MarkdownUtils.parse = jest.fn()
    MarkdownUtils.parse.mockReturnValue(content)
    MarkdownUtils.slidesCount = jest.fn()
    MarkdownUtils.slidesCount.mockReturnValue(3)

    SlideUtils.getSlide = jest.fn()
    SlideUtils.getSlide.mockReturnValue(<Slide content={content[0]} onClick={() => null} />)

    RouterUtils.getSlideIndexFromProps = jest.fn()
    RouterUtils.getSlideIndexFromProps.mockReturnValue(1)

    props = { navigateTo: () => null }
    tree = shallow(<SlidesContainer {...props} />)
  })

  afterEach(() => {
    MarkdownUtils.parse.mockReset()
    MarkdownUtils.slidesCount.mockReset()
    SlideUtils.getSlide.mockReset()
    RouterUtils.getSlideIndexFromProps.mockReset()
  })

  test('renders correctly', () => {
    expect(tree).toMatchSnapshot()
  })
})
