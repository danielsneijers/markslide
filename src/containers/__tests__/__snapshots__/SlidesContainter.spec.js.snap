// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/SlidesContainer renders correctly 1`] = `
<div>
  <ProgressBar
    offset={66.66666666666667}
  />
  <Slide
    content="<h1>Nice title</h1>
  "
    onClick={[Function]}
  />
</div>
`;
